src/MallardMessageHandlers/AuthenticationToken/AuthenticationTokenHandler.cs

```diff
@@ -177,25 +181,33 @@ TAuthenticationToken unauthorizedToken
 			{
 				_logger.LogWarning($"The access tokens are different. No need to refresh, returning the current token '{currentToken}'.");
 
+				// Release the semaphore.
+				_semaphore.Release();
+
 				return currentToken;
 			}
 
+			var refreshedToken = default(TAuthenticationToken);
+
 			try
 			{
 				_logger.LogDebug($"Refreshing token: '{unauthorizedToken}'.");
 
-				var refreshedToken = await _tokenProvider.RefreshToken(ct, request, currentToken);
+				refreshedToken = await _tokenProvider.RefreshToken(ct, request, currentToken);
 
 				_logger.LogInformation($"Refreshed token: '{unauthorizedToken}' to '{refreshedToken}'.");
-
-				return refreshedToken;
 			}
 			catch (Exception e)
 			{
 				_logger.LogError(e, $"Failed to refresh token: '{unauthorizedToken}'.");
-
-				return default;
 			}
+			finally
+			{
+				// Release the semaphore.
+				_semaphore.Release();
```

Comment:
I think you should keep only this one.

Reply:
I didn't have a choice to put it everywhere to make sure that I didn't miss one. After thinking about it, I instead extracted the get authentication token into a seperate method so we can isolate the wait and release of the semaphore into one place only